lms/djangoapps/ccx/modulestore.py

```diff
@@ -0,0 +1,282 @@
+# -*- coding: utf-8 -*-
+"""A modulestore wrapper
+
+It will 'unwrap' ccx keys on the way in and re-wrap them on the way out
+
+In practical terms this means that when an object is retrieved from modulestore
+using a CCXLocator or CCXBlockUsageLocator as the key, the equivalent
+CourseLocator or BlockUsageLocator will actually be used. And all objects
+returned from the modulestore will have their keys updated to be the CCX
+version that was passed in.
+"""
+from contextlib import contextmanager
+from functools import partial
+from ccx_keys.locator import CCXLocator, CCXBlockUsageLocator
+from opaque_keys.edx.locator import CourseLocator, BlockUsageLocator
+
+
+def strip_ccx(val):
+    retval = val
+    ccx_id = None
+    if isinstance(retval, CCXLocator):
+        ccx_id = retval.ccx
+        retval = retval.to_course_locator()
+    elif isinstance(object, CCXBlockUsageLocator):
+        ccx_id = retval.course_key.ccx
+        retval = retval.to_block_locator()
+    elif hasattr(retval, 'location'):
+        retval.location, ccx_id = strip_ccx(retval.location)
+    return retval, ccx_id
+
+
+def restore_ccx(val, ccx_id):
+    if isinstance(val, CourseLocator):
+        return CCXLocator.from_course_locator(val, ccx_id)
+    elif isinstance(val, BlockUsageLocator):
+        ccx_key = restore_ccx(val.course_key, ccx_id)
+        val = CCXBlockUsageLocator(ccx_key, val.block_type, val.block_id)
+    if hasattr(val, 'location'):
+        val.location = restore_ccx(val.location, ccx_id)
+    if hasattr(val, 'children'):
+        val.children = restore_ccx_collection(val.children, ccx_id)
+    return val
+
+
+def restore_ccx_collection(field_value, ccx_id=None):
+    if ccx_id is None:
+        return field_value
+    if isinstance(field_value, list):
+        field_value = [restore_ccx(fv, ccx_id) for fv in field_value]
+    elif isinstance(field_value, dict):
+        for key, val in field_value.iteritems():
+            field_value[key] = restore_ccx(val, ccx_id)
+    else:
+        field_value = restore_ccx(field_value, ccx_id)
+    return field_value
+
+
+@contextmanager
+def remove_ccx(to_strip):
+    stripped, ccx = strip_ccx(to_strip)
+    yield stripped, partial(restore_ccx_collection, ccx_id=ccx)
+
+
+class CCXModulestoreWrapper(object):
+
+    def __init__(self, modulestore):
+        self._modulestore = modulestore
+
+    def __getattr__(self, name):
+        """pass missing attributes through to _modulestore
+        """
+        return getattr(self._modulestore, name)
+
+    def _clean_locator_for_mapping(self, locator):
+        with remove_ccx(locator) as stripped:
+            locator, restore = stripped
+            retval = self._modulestore._clean_locator_for_mapping(locator)
+            return restore(retval)
```

Comment:
And then this could just be: `return restore(self._modulestore._clean_locator_for_mapping(locator))`